Show reasons for disabled service in notification dialog (fixes #1263)

capi opened the pull request:
This implementation aims to change the actual run-condition check as little as possible, but returns a set of reasons instead of a single boolean. Since the original implementation returns early in several cases, this early returns are simulated but still attempt to collect all applying reasons.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java:
```diff
@@ -142,19 +140,27 @@ public void onReceive(Context context, Intent intent) {
      * Determines if Syncthing should currently run.
      */
-    private boolean decideShouldRun() {
+    private Set<String> decideShouldRun() {
```
Let's create a auxilary class / DTO / POJO whatever the hell, and potentially add reasons for running and a boolean?
I guess reasons can be both positive and negative, so we could store both and imply that the top one is the reason for our current state, which then allows people to understand the order of positives and negatives.

capi replied:
I considered this the first time, but since the run logic had some issues previously I just didn't want to change too much to it. At the moment it aims to just be the simplest solution to the request without actually changing the logic. Future improvements like you said would be sensible, I just didn't want to do it as first step.

AudriusButkevicius replied:
Well the DTO would only have the stack of messages, that doesn't change the logic, and I think that's still valuable to understand where the issues are (when reading the stack of messages people post)

capi commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -258,7 +263,20 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+    private void onUpdatedShouldRunDecision(Set<String> reasons) {
+        boolean newShouldRunDecision = reasons.isEmpty();
+        boolean reasonsChanged;
+        synchronized (mStateLock) {
```
True, it should be `mCurrentState` thanks for catching that.

capi commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -404,6 +404,18 @@ Bitte melden Sie auftretende Probleme via GitHub.</string>
   <string name="syncthing_disabled_change_settings">Einstellungen ändern</string>
   <!--Button text on the "syncthing disabled" dialog, used as menu item to stop syncthing service if "always_run_in_background" is true-->
   <string name="exit">Beenden</string>
+  <!-- Reasons for "syncthing disabled" dialog -->
```
Yeah, I was afraid of something like that. I checked the history of the file and saw manual commits outside of transifex so I hoped it would be OK. I'll simply remove the file from the push and see that it'll be translated on transifex instead.

capi commented:
I've changed the changes requested by @Catfriend1. @AudriusButkevicius is right about the DTO, but I'll need some more time for this, hopefully later tonight. Using an immutable DTO would also remove the need for all the synchronization and make the code clearer, so yes, I'll change it to that.

capi commented on app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java:
```diff
@@ -166,18 +172,23 @@ private boolean decideShouldRun() {
         boolean prefRespectPowerSaving = mPreferences.getBoolean(Constants.PREF_RESPECT_BATTERY_SAVING, true);
         boolean prefRespectMasterSync = mPreferences.getBoolean(Constants.PREF_RESPECT_MASTER_SYNC, false);
 
+        Set<String> blockerReasons = new TreeSet<>(); // TreeSet so that orders are sorted
```
Yes, in the DTO that was requested I'll do exactly this.

capi commented on app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java:
```diff
@@ -166,18 +172,23 @@ private boolean decideShouldRun() {
         boolean prefRespectMasterSync = mPreferences.getBoolean(Constants.PREF_RESPECT_MASTER_SYNC, false);
 
+        Set<String> blockerReasons = new TreeSet<>(); // TreeSet so that orders are sorted
+        Boolean resultForced = null; // need to remember if we encountered a return in previous version
```
The final result is a `boolean`, the problem is, that the initial code returns at a few stages, so I can set the final result only, if no other return would have returned previously. Boolean allows me to check if there would have been already a return, and only set the final result if the existing code wouldn't have already returned.

Nutomic replied:
Yeah but putting three values in a boolean like this seems really confusing for me.

capi replied:
I hope you agree with my new solution that avoids using this at all, together with a DTO.

capi commented:
So, I now force-pushed the draft that uses an enum and a DTO for the reasons. I'd appreciate feedback.

Nutomic commented on app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java:
```diff
@@ -16,19 +16,31 @@
+import static com.nutomic.syncthingandroid.model.RunConditionCheckResult.Reason.ON_CHARGER;
+import static com.nutomic.syncthingandroid.model.RunConditionCheckResult.Reason.POWERSAVING_ENABLED;
+import static com.nutomic.syncthingandroid.model.RunConditionCheckResult.Reason.WIFI_SSID_NOT_WHITELISTED;
+import static com.nutomic.syncthingandroid.model.RunConditionCheckResult.Reason.WIFI_WIFI_IS_METERED;
```
I think you can import the enum as a whole, and not list every item separately? Not sure, my Java is a bit rusty.

Otherwise looks good 👍

capi replied:
Could be, it was generated that way by Android Studio. I'll check.

capi commented:
I suppose the failed TeamCity build is due to the changes of the Go version in the docker container?

Catfriend1 commented on app/src/main/java/com/nutomic/syncthingandroid/model/RunConditionCheckResult.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.Collections;
+import java.util.List;
+
+public class RunConditionCheckResult {
```
This new DTO is very nice, good work :-)

capi commented on app/src/main/java/com/nutomic/syncthingandroid/model/RunConditionCheckResult.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private final boolean mShouldRun;
+    private List<Reason> mBlockReasons;
```
This should be `final` as well.

capi commented:
I've added the final remarks of @Catfriend1 and @Nutomic.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -578,6 +595,30 @@ private void onServiceStateChange(State newState) {
+                if (listener != null) {
+                    listener.onRunConditionCheckResultChanged(result);
+                } else {
+                    i.remove();
```
Why remove here?

capi replied:
To be honest, I just copied the existing method and did not think too much about it.

Catfriend1 replied:
If an activity that registered a listener is unavailable being set to null, it's automatically cleant up. Nutomic's code has this all over the place.

capi replied:
I've not addressed this one, since it's consistent with the other code. If it really is a blocker, I can of course adjust.

AudriusButkevicius replied:
I understand the value gets gc'ed and becomes null, but how given there is a pointer in this set, which should prevent collection?

If we had a weakref, I'd understand it, but we don't have a weakref, we have a strong ref.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -262,7 +269,17 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+    private void onUpdatedShouldRunDecision(RunConditionCheckResult result) {
+        boolean newShouldRunDecision = result.isShouldRun();
+        boolean reasonsChanged;
+        synchronized (mCurrentCheckResult) {
```
Why do you need to synchronize on an atomic? That makes no sense.

capi replied:
Because I want consistency over the compare and assignment. I was thinking the same, but otherwise the assignment will race.

capi replied:
Fixed by your suggestion of using `getAndSet()` instead.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -262,7 +269,17 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+        boolean reasonsChanged;
+        synchronized (mCurrentCheckResult) {
+            reasonsChanged = !mCurrentCheckResult.get().equals(result);
+            mCurrentCheckResult.set(result);
```
```
changed = !mCurrentCheckResult.getAndSet(result).equals(result)
```

capi replied:
Yes, but in this case we assign the result even if not changed. On the other hand since it is immutable, it doesn't hurt.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/model/RunConditionCheckResult.java:
```diff
@@ -0,0 +1,68 @@
+
+    public RunConditionCheckResult() {
+        mBlockReasons = Collections.emptyList();
+        mShouldRun = true;
```
`this(Collections.emptyList())`?

capi replied:
Will do.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/model/RunConditionCheckResult.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public int hashCode() {
+        return mShouldRun ? 17 : 42;
```
why not Object.hashCode(mShouldRun) or something like that?

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/model/RunConditionCheckResult.java:
```diff
@@ -0,0 +1,68 @@
+        RunConditionCheckResult that = (RunConditionCheckResult) o;
+        return mShouldRun == that.mShouldRun && mBlockReasons.size() == that.mBlockReasons.size()
+                && mBlockReasons.containsAll(that.mBlockReasons);
+    }
```
IntelliJ (and potentially Android studio) can generate both hashcode and equals whcih has a better implementation.
It seems strange that equality takes reasons into account, and hash code does not.

capi replied:
You are looking at the old version. This is already outdated by my latest push. This version has been replaced by the InteliJ version.

AudriusButkevicius replied:
```
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        RunConditionCheckResult that = (RunConditionCheckResult) o;

        if (mShouldRun != that.mShouldRun) return false;
        return mBlockReasons.equals(that.mBlockReasons);
    }

    @Override
    public int hashCode() {
        int result = (mShouldRun ? 1 : 0);
        result = 31 * result + mBlockReasons.hashCode();
        return result;
    }
```

is what I see. Usually IntelliJ generates `Object.equals(x, y) && ...` etc, and `Object.hash(o) + Object.hash(y)`

capi replied:
I needed to the revert to the default generated version (you can change that in the generate dialog), since our target API level does not have `Object.hash(o)` available. This is available from API 19, we are still targeting API 16.

capi commented:
I'm skiing for a week, once I'm back I'll address the remaining points. Thanks for taking all the time for the review.

capi commented:
I've rebased to latest master and I've addressed the reviews of @AudriusButkevicius. I reworked the `hashCode()` implementation (the last one was fulfilling the contract, but I agree that it looked strange). I still need to keep my own implementations since the generated versions don't do a deep compare of the list.
Also I made the empty constructor private and introduced a `SHOULD_RUN` static final field, since using this makes it more readable in my opinion.
Also I removed the synchronization of the `AtomicReference` by using the suggested `getAndSet()` approach.

AudriusButkevicius commented:
I am still not sure I understand why we need magical hash and equals, why we can't use IDE generated ones.

capi commented:
The one it did for me, it did not. And I also just noticed that I broke it (while me last version was not optimal but technically correct). So **PLEASE DO NOT MERGE** until I fixed it.

I'll double-check tonight, but last time I let it generate equals and hash-code it used equals on the list, which does not do a deep comparison of the list content. And honestly I think it can't, because in this case the hash would not be stable as soon as the list changed. You can only do a deep compare and use it for hash if the list itself is immutable.

AudriusButkevicius commented:
Right, I guess if the order should be ignored, the equals should do a set comparison?

At which point I guess you do need a custom equals and hashcode.

capi commented:
@AudriusButkevicius: I have thought about it, you are right. Since I explicitly use a `List` and not a `Set`, I should not suddenly use `Set` semantics for `equals()` and `hashCode()`, therefore I now use the InteliJ-generated implementation.
